crypto-ffi/src/generic/mod.rs

```diff
@@ -189,52 +192,151 @@ pub enum CoreCryptoError {
+        ///
+        /// Hopefully only ever use this in conjunction with `interior_matches!`, because for most sane
+        /// circumstances, `if let` is the better design pattern.
+        macro_rules! matches_option {
+            ($val:expr, $pattern:pat $(if $guard:expr)? => $out:expr) => {
+                match ($val) {
+                    $pattern $(if $guard)? => Some($out),
+                    _ => None,
+                }
+            };
+        }
+
+        /// This is moderately horrific and we hopefully will not require it anywhere else, but
+        /// it solves a real problem here: how do we match against the innermost error variants,
+        /// when we have a heterogenous set of types to match against?
+        macro_rules! match_heterogenous {
+            ($err:expr => {
+                $( $pattern:pat $(if $guard:expr)? => $var:expr, )*
+                ||=> $default:expr,
+            }) => {{
+                if false {unreachable!()}
+                $(
+                    else if let Some(v) = matches_option!($err.downcast_ref(), Some($pattern) $(if $guard)? => $var) {
+                        v
+                    }
+                )*
+                else {
+                    $default
+                }
+            }};
+        }
+
+        match_heterogenous!(innermost => {
+            core_crypto::LeafError::ConversationAlreadyExists(id) => MlsError::ConversationAlreadyExists(id.clone()).into(),
+            core_crypto::mls::conversation::Error::BufferedFutureMessage => MlsError::BufferedFutureMessage.into(),
+            core_crypto::mls::conversation::Error::DuplicateMessage => MlsError::DuplicateMessage.into(),
+            core_crypto::mls::conversation::Error::MessageEpochTooOld => MlsError::MessageEpochTooOld.into(),
+            core_crypto::mls::conversation::Error::SelfCommitIgnored => MlsError::SelfCommitIgnored.into(),
+            core_crypto::mls::conversation::Error::StaleCommit => MlsError::StaleCommit.into(),
+            core_crypto::mls::conversation::Error::StaleProposal => MlsError::StaleProposal.into(),
+            core_crypto::mls::conversation::Error::UnbufferedFarFutureMessage => MlsError::WrongEpoch.into(),
+            core_crypto::mls::conversation::Error::OrphanWelcome => MlsError::OrphanWelcome.into(),
+            core_crypto::mls::Error::UnmergedPendingGroup => MlsError::UnmergedPendingGroup.into(),
+            ||=> MlsError::Other(error.innermost_error_message()).into(),
+        })
+    }
 }
 
 // This implementation is intended to be temporary; we're going to be completely restructuring the way we handle
 // errors in `core-crypto` soon. We can replace this with better error patterns when we do.
 //
 // Certain error mappings could apply to both MLS and Proteus. In all such cases, we map them to the MLS variant.
 // When we redesign the errors in `core-crypto`, these ambiguities should disappear anyway.
```

Comment:
Does this comment still apply?

Reply:
Hmm. Not as it was originally written, when it referred to this work. But now I'm hopeful that we can come up with a Big Refactor Plan in a month which leads to us getting better error-handling patterns in the future. So I guess I would still leave it in there for now.